ports/stm32f4/common-hal/pulseio/PulseIn.c

```diff
@@ -34,46 +34,296 @@
+        GPIO_InitStruct.Mode = GPIO_MODE_IT_RISING_FALLING;
+        GPIO_InitStruct.Pull = GPIO_NOPULL;
+        HAL_GPIO_Init(pin_port(self->pin->port), &GPIO_InitStruct);
+    }
+
+    self->first_edge = true;
+    self->paused = false;
+    self->last_ms = 0;
+    self->last_us = 0;
+
+    HAL_NVIC_EnableIRQ(self->irq);
 }
 
 void common_hal_pulseio_pulsein_clear(pulseio_pulsein_obj_t* self) {
+    HAL_NVIC_DisableIRQ(self->irq);
+    self->start = 0;
+    self->len = 0;
+    HAL_NVIC_EnableIRQ(self->irq);
 }
 
 uint16_t common_hal_pulseio_pulsein_get_item(pulseio_pulsein_obj_t* self, int16_t index) {
-    return 0;
+    HAL_NVIC_DisableIRQ(self->irq);
+    if (index < 0) {
+        index += self->len;
+    }
+    if (index < 0 || index >= self->len) {
+        HAL_NVIC_EnableIRQ(self->irq);
+        mp_raise_IndexError(translate("index out of range"));
+    }
+    uint16_t value = self->buffer[(self->start + index) % self->maxlen];
+    HAL_NVIC_EnableIRQ(self->irq);
+    return value;
 }
 
 uint16_t common_hal_pulseio_pulsein_popleft(pulseio_pulsein_obj_t* self) {
-    return 0;
+    if (self->len == 0) {
+        mp_raise_IndexError(translate("pop from an empty PulseIn"));
+    }
+    HAL_NVIC_DisableIRQ(self->irq);
+    uint16_t value = self->buffer[self->start];
+    self->start = (self->start + 1) % self->maxlen;
+    self->len--;
+    HAL_NVIC_EnableIRQ(self->irq);
+
+    return value;
 }
 
 uint16_t common_hal_pulseio_pulsein_get_maxlen(pulseio_pulsein_obj_t* self) {
-    return 0;
+    return self->maxlen;
 }
 
 bool common_hal_pulseio_pulsein_get_paused(pulseio_pulsein_obj_t* self) {
-    return 0;
+    return self->paused;
 }
 
 uint16_t common_hal_pulseio_pulsein_get_len(pulseio_pulsein_obj_t* self) {
-    return 0;
+    return self->len;
+}
+
+static void assign_EXTI_Interrupt(pulseio_pulsein_obj_t* self, uint8_t num) {
+    if (num == 0) {
+        self->irq = EXTI0_IRQn;
+    } else if (num == 1) {
+        self->irq = EXTI1_IRQn;
+    } else if (num == 2) {
+        self->irq = EXTI2_IRQn;
+    } else if (num == 3) {
+        self->irq = EXTI3_IRQn;
+    } else if (num == 4) {
+        self->irq = EXTI4_IRQn;
+    } else if (num >= 5 && num <= 9 ) {
+        self->irq = EXTI9_5_IRQn;
+    } else if (num >= 10 && num <= 15) {
+        self->irq = EXTI15_10_IRQn;
+    }
+}
+
+void EXTI0_IRQHandler(void)
+{
+    EXTI->PR = 1 << 0;
+    pulsein_handler(0);
+}
+void EXTI1_IRQHandler(void)
+{
+    EXTI->PR = 1 << 1;
+    pulsein_handler(1);
+}
+void EXTI2_IRQHandler(void)
+{
+    EXTI->PR = 1 << 2;
+    pulsein_handler(2);
+}
+void EXTI3_IRQHandler(void)
+{
+    EXTI->PR = 1 << 3;
+    pulsein_handler(3);
+}
+void EXTI4_IRQHandler(void)
+{
+    EXTI->PR = 1 << 4;
+    pulsein_handler(4);
+}
+void EXTI9_5_IRQHandler(void)
+{
+    uint32_t pending = EXTI->PR;
+    if(pending & (1 << 5)) {
+        EXTI->PR = 1 << 5;
+        pulsein_handler(5);
+    } else if (pending & (1 << 6)) {
```

Comment:
Don't do `else if` here and below because more than one pin may actually have interrupted. Getting them all in the same handler will save the overhead of re-calling it.